src/machine/machine_atmega.go

```diff
@@ -5,11 +5,120 @@ package machine
 
 import (
 	"device/avr"
+	"errors"
 	"runtime/interrupt"
 	"runtime/volatile"
 	"unsafe"
 )
 
+// EECR bitfield values.
+// The constants provided by the AVR package are not correct, so these are the correct fields instead.
+const (
+	eecrEERE  = avr.EECR_EERE - 1
+	eecrEEPE  = avr.EECR_EEPE - 1
+	eecrEEMPE = avr.EECR_EEMPE - 2
+	eecrEERIE = avr.EECR_EERIE - 5
+	eecrEEPM0 = avr.EECR_EEPM0 - 12
+	eecrEEPM1 = avr.EECR_EEPM1 - 27
+)
+
+// EEPROM on ATMega
+type EEPROM struct {
+}
+
+var EEPROM0 = EEPROM{}
+
+// setAddress sets the address for a given read or write into the MCUs EEARH/L register.
+func (e EEPROM) setAddress(addr int64) error {
+	if addr < 0 || addr > e.Size() {
+		return errors.New("address provided is out of bounds")
+	}
+
+	avr.EEARH.Set(uint8(addr >> 8))
+	avr.EEARL.Set(uint8(addr & 0xFF))
+
+	return nil
+}
+
+// WriteAt writes len(data) bytes in the EEPROMs at the provided offset.
+func (e EEPROM) WriteAt(data []byte, off int64) (int, error) {
+	written := 0
+	for i, value := range data {
+		if err := e.WriteByteAt(value, off+int64(i)); err != nil {
+			return written, err
+		}
+		written++
+	}
+	return written, nil
+}
+
+// WriteByteAt performs the logic to writes a byte into the EEPROM at the given address.
+func (e EEPROM) WriteByteAt(value byte, addr int64) error {
+	notReady := true
+	for notReady {
+		// Wait for the completion of the previous write (if there is one)
+		if !avr.EECR.HasBits(1 << eecrEEPE) {
```

Comment:
```suggestion
		if !avr.EECR.HasBits(avr.EECR_EEPE) {
```

But actually this whole loop can be simplified to the following:

```go
for avr.EECR.HasBits(avr.EECR_EEPE) {}
```